// protocol-slip/dist/encode.js
"use strict";
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
exports.encodeMessage = (data) => {
    let remainingPacket = data;
    let endPosition = remainingPacket.indexOf(constants_1.END);
    let escPosition = remainingPacket.indexOf(constants_1.ESC);
    const parts = [];
    while (endPosition > -1 || escPosition > -1) {
        const nextSlice = Math.min(endPosition === -1 ? Infinity : endPosition, escPosition === -1 ? Infinity : escPosition);
        const part = remainingPacket.slice(0, nextSlice);
        const escapeSeq = nextSlice === endPosition ? constants_1.END_ESC_SEQ : constants_1.ESC_ESC_SEQ;
        parts.push(part, escapeSeq);
        remainingPacket = remainingPacket.slice(nextSlice + 1);
        // find the next thing to escape
        if (nextSlice === endPosition) {
            endPosition = remainingPacket.indexOf(constants_1.END);
            escPosition = escPosition === -1 ? -1 : escPosition - (part.length + 1);
        }
        else {
            endPosition = endPosition === -1 ? -1 : endPosition - (part.length + 1);
            escPosition = remainingPacket.indexOf(constants_1.ESC);
        }
    }
    parts.push(remainingPacket, constants_1.END);
    return Buffer.concat(parts);
};
function _asyncEncode(iterable) {
    return __asyncGenerator(this, arguments, function* _asyncEncode_1() {
        var e_1, _a;
        try {
            for (var iterable_1 = __asyncValues(iterable), iterable_1_1; iterable_1_1 = yield __await(iterable_1.next()), !iterable_1_1.done;) {
                const data = iterable_1_1.value;
                yield yield __await(exports.encodeMessage(data));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (iterable_1_1 && !iterable_1_1.done && (_a = iterable_1.return)) yield __await(_a.call(iterable_1));
            }
            finally { if (e_1) throw e_1.error; }
        }
    });
}
function* _syncEncode(iterable) {
    for (const data of iterable) {
        yield exports.encodeMessage(data);
    }
}
function encode(iterable) {
    if (iterable[Symbol.asyncIterator]) {
        return _asyncEncode(iterable);
    }
    return _syncEncode(iterable);
}
exports.encode = encode;
